working/base-interface-final/feature-specification.md

```diff
@@ -0,0 +1,511 @@
+    extended.
+
+*   **Trust but verify.** In the earlier example, it's not clear what the author
+    *intends*. Maybe they deliberately didn't put any modifiers on `MySubclass`
+    because they *want* to re-add the capability that its superclass removed.
+    But maybe they just didn't notice that `NoExtend` removed them, or they
+    forgot to put `interface` on `MySubclass`.
+
+    Since it's not clear what they meant, the language could require them to
+    clarify. If you define a subtype of a type that has removed a capability, we
+    could require you to annotate specifically when you re-add that capability.
+    If you don't intend to re-add a capability, you restate the restriction:
+
+    ```dart
+    interface class NoExtend {}
+    interface class MySubclass extends NoExtend {}
+    ```
+
+    And if you *do* intend to loosen it, you make that explicit by some marker
+    like:
+
+    ```dart
+    interface class NoExtend {}
+    reopen class MySubclass extends NoExtend {}
+    ```
+
+    Here "reopen" means, "I know I didn't put any other modifier here and that
+    means this class has more capabilities than my parent."
+
+    Personally, I think this is probably more modifiers than we want and is
+    more trouble than it's worth. I worry about having to explain to users that
+    a class marked `reopen` means the same thing as a class not marked with it.
+    But I do think it could be useful to offer this as a lint with a metadata
+    annotation for users that are more cautious, like:
+
+    ```dart
+    interface class NoExtend {}
+    @reopen class MySubclass extends NoExtend {}
+    ```
+
+This proposal takes the last option where types have exactly the restrictions
+they declare but a lint can be turned on for users who want to be reminded if
+they re-add a capability in a subtype.
+
+Finally, to the actual proposal...
+
+## Syntax
+
+This proposal builds on the existing sealed types proposal so the grammar
+includes those changes. The full set of modifiers that can appear before a class
+or mixin are `abstract`, `sealed`, `base`, `interface`, and `final`. Some
+combinations don't make sense:
+
+*   `sealed` implies `abstract`, so they can't be combined.
+*   `sealed` implies non-extensibility, so can't be combined with `interface`
+    or `final`.
+*   `sealed` implies non-implementability, so can't be combined with `base` or
+    `final`.
+*   `base`, `interface`, and `final` all control the same two capabilities so
+    are mutually exclusive.
+*   `final` and `interface` imply that the type can't be used as a superclass,
+    but the reason for defining a mixin is so that it can be mixed in and become
+    a superclass, so we don't allow those on mixins.
+
+The remaining valid combinations are:
+
+```
+class
+sealed class
+base class
+interface class
+final class
+abstract class
+abstract base class
+abstract interface class
+abstract final class
```

Comment:
> it's not obvious to me that `sealed` needs to differ from `abstract final`.

With `abstract final` I can still add a new subclass in the same library at any time and it doesn't break consumers. With `sealed` I need to promise that I will never add another direct subclass after I publish.

Reply:
> However, `sealed class` and `abstract final class` is the same thing according to these bits, and it's not obvious to me that `sealed` needs to differ from `abstract final`.

There's an extra fourth bit: exhaustiveness checks. Marking a class `sealed` opts that class into exhaustiveness checks in switches. Marking it `abstract final class` does not:

```dart
sealed class A {}
class A2 extends A {}
class A3 extends A {}

abstract final class B {}
class B2 extends B {}
class B3 extends B {}

test(A a, B b) {
  print(switch (a) {
    A1() => 'A1',
    A2() => 'A2'
  }); // No error here.

  print(switch (b) {
    B1() => 'B1',
    B2() => 'B2'
  }); // Error, switch isn't (known to be) exhaustive.
}
```

In principle, yes, the compiler *could* tell that switches over subclasses of `abstract final` classes are not exhaustive. But using `abstract final` instead of `sealed` explicitly opts *out* of that checking. That forces users to write:

```dart
test(B b) {
  print(switch (b) {
    B1() => 'B1',
    B2() => 'B2',
    _ => 'some default clause'
  });
}
```

That in turn means that if the author of `B` later adds another subclass, then the switch (which may be in a different library or package) still compiles without error. You can think of `abstract final` as sort of `sealed late` in that it turns exhaustiveness compile time errors into runtime errors (or at least runtime *handled*) by forcing users to add default cases to their switches.

It's a subtle distinction, but one that could be useful for some package maintainers who want the ability to ship new subclasses of a class without it being a breaking change to users of the API.

Reply:
Makes sense!